esp_custom_part_ota/src/esp_custom_part_ota.c

```diff
@@ -0,0 +1,281 @@
+            size_t data_size = MIN(block_size, handle->backup_len - copy_size);
+            ret = esp_partition_read(handle->update_partition, copy_size, data, data_size);
+            if (ret != ESP_OK) {
+                goto end;
+            }
+            ret = esp_partition_write(handle->backup_partition, copy_size, data, data_size);
+            if (ret != ESP_OK) {
+                goto end;
+            }
+            copy_size += data_size;
+        }
+    }
+end:
+    free(data);
+    if (ret == ESP_OK) {
+        ret = esp_partition_erase_range(handle->update_partition, 0, handle->update_partition->size);
+        if (ret == ESP_OK) {
+            handle->need_erase = 0;
+            ESP_LOGI(TAG, "Successfully erased update partition");
+        }
+    }
+
+    // Store backup data information in the default NVS partition.
+    nvs_handle_t backup_info_handle = 0;
+    ret = nvs_open(BACKUP_STORAGE_NAMESPACE, NVS_READWRITE, &backup_info_handle);
+    if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "Failed to store backup information: %s", esp_err_to_name(ret));
+        return ret;
+    }
+    ret = nvs_set_u32(backup_info_handle, "backup_len", handle->backup_len);
+    if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "Failed to store backup information: %s", esp_err_to_name(ret));
+        return ret;
+    }
+    ret = nvs_commit(backup_info_handle);
+    if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "Failed to store backup information: %s", esp_err_to_name(ret));
+        return ret;
+    }
+    nvs_close(backup_info_handle);
+
+    return ret;
+}
+
+esp_err_t esp_custom_part_ota_partition_restore(esp_custom_part_ota_handle_t ctx)
+{
+    if (!ctx) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_custom_part_ota_t *handle = (esp_custom_part_ota_t *)ctx;
+    if (!handle->backup_partition || !handle->update_partition) {
+        ESP_LOGE(TAG, "Partition(s) not set. Cannot restore");
+        return ESP_FAIL;
+    }
+    esp_err_t ret = ESP_FAIL;
+
+    // Check if valid backup is present in the backup partition by reading backup information from the default NVS partition.
+    if (handle->backup_len == 0) {
+        nvs_handle_t backup_info_handle = 0;
+        ret = nvs_open(BACKUP_STORAGE_NAMESPACE, NVS_READWRITE, &backup_info_handle);
+        if (ret != ESP_OK) {
+            ESP_LOGE(TAG, "Failed to fetch backup information from NVS: %s", esp_err_to_name(ret));
+            return ret;
+        }
+        ret = nvs_get_u32(backup_info_handle, "backup_len", (uint32_t *)(&(handle->backup_len)));
+        if (ret != ESP_OK) {
+            ESP_LOGE(TAG, "Failed to fetch backup information from NVS: %s", esp_err_to_name(ret));
+            return ret;
+        }
+        nvs_close(backup_info_handle);
+    }
+    if (handle->backup_len == 0) {
+        ESP_LOGI(TAG, "No backup present in the backup partition. Nothing to restore");
+        return ESP_OK;
+    }
+    if (handle->backup_len > 0) {
```

Comment:
Could there be a scenario where this length is partial and not for full image? Do we need any image integrity checks (e.g., SHA sum) in the NVS?

Reply:
Yes, the user can set the length of the data to be backed-up. I don't think we need any image integrity checks. We store the size of data in NVS. Also in case of any error, the APIs will return relevant error code.